### DISCUSSION TITLE
USB Programmer for the SMT32F4 and building a bootloader

### DISCUSSION THREAD

**Discussion by mrbib** [mriscoc/Ender3V2S1#954] [General]
First, has anyone purchased a USB programmer interface board that can mate with the SMT32F4 boards in the S1 models? I think it's a 4-pin header next to the touchscreen connector port? Which one? Any problems?

Second, has anyone built the bootloader for the SMT32F4 board using PlatformIO and Visual Studio? 

Why do I ask? I'd like to replace the 256k SoC on my board with the 512k chip. When I do that, I'll need to replace the bootloader before I can put firmware on it.

**Comment by mriscoc** [comment]
Seems there is a boot loader compatible with STM32F4 RET here: https://github.com/CrealityOfficial/E3-Free-runs-Silent-Motherboard

**Reply by Shaische** [reply]
@mriscoc If you don't mind helping, what address do you flash the boot loader and firmware respectively? I flashed boot loader at 0x08000000 and verified the flash successfully. Then flashed firmware at 0x08010000 and verified successfully via STM32 Cube Programmer. But when I go back to verify boot loader again, I get: "Error: Data mismatch found at address  0x08000000 (byte = 0x00 instead of 0x88)"

I'm trying to understand what I'm doing wrong here because I'm unable to connect via USB, it doesn't show up as a port, which tells me it's likely that the firmware isn't running to create the port? Any help appreciated as I'm new to STM32 SWD programming!

**Reply by mriscoc** [reply]
For reference, when you flash the firmware via ST-link, the bootloader could be erased, so, flash the bootloader via ST-link and then the firmware using an SD-card like a normal procedure.

**Comment by Crazy-Charles** [comment]
The STLINK V2 programmer is the one i use, i also replaced my 401RC with the RE version and at the time the boot loader i extracted from the RC didnt work on the RE, it would program from SD ok but only on the first time as it didnt erase properly until i patched it to erase all blocks.
Not tried the ones that are available now as i did my "upgrade" a long time ago!

**Reply by mrbib** [reply]
Want to tell us more about your process?

I'm mostly interested in doing this because the 512K part is under $10, plus the time to unsolder/resolder with the new part.

**Reply by Crazy-Charles** [reply]
I upgraded one of my Ender 3 S1 boards which originally had the RC, so a processor swap was all that was required.
The STLINK V2 i got is a chinese copy of the original and works with the STM32Cube programmer e.g. https://www.ebay.co.uk/itm/394652149030

Once connected to the 4 pin header (pin assignments are silk screened on the board) you may have to short the mcu reset pin to ground at power on for the programmer to connect and program the bootloader.

**Reply by mrbib** [reply]
I'm looking at this programmer interface on Amazon (similar on eBay): https://www.amazon.com/HiLetgo-Emulator-Downloader-Programmer-STM32F103C8T6/dp/B07SQV6VLZ/

Do you agree that there are pins to line up with the header on the board? (Photo from the super-high resolution one on Reddit, https://www.reddit.com/r/Ender3S1/comments/tfaqv2/any_idea_what_motherboard_this_is_ender_3s1_has/ . My own photo wasn't this clear.)

![image](https://github.com/mriscoc/Ender3V2S1/assets/77157402/c0f63f9c-f099-4eae-80ec-87ee80b9575c)

**Reply by Crazy-Charles** [reply]
From left to right on your image...
GND - SWCLK - SWDIO - 3.3v

I do have a couple of those type of programmers, you can only use them with the old outdated ST-LINK software and TBH i found them troublesome and unreliable but they will work. Just don't try and update the onboard firmware, it WILL fail unlike the clone i linked to.

The square unpopulated pin hole in the board is the power on MCU reset pin, the round one is ground.

**Reply by mrbib** [reply]
SO you prefer the white box interface. Good to know.

Once the bootloader is in place, I'd just want to use the regular SD card update method, though.

**Reply by Crazy-Charles** [reply]
Good luck with the transplant, BTW the board will also take the F103RE, just remove cap C742 as that is the ONLY difference between the two boards.

**Reply by mrbib** [reply]
@Crazy-Charles I found a real ST-Link (not a clone) for $15 shipped. Should be here by Friday. :)

**Comment by gitfve** [comment]
-It will be so nice if you post back what you find out. And some hint on how you have done the Boot Loader programming....
I have a "Ender3 S1 Pro" with the STM32F401RCT6 (ROM:256KB RAM:64KB)
Board version: CR-FDM-v24S1-301 (CR4NS200141C13)
And want to change the cpu to the RE with more ROM and RAM.

**Comment by gitfve** [comment]
Any progress on this?

**Reply by Crazy-Charles** [reply]
@gitfve What info are you looking for exactly?

**Reply by gitfve** [reply]
As my board is with the 256Kbyte chip i think to replace it with the 512Kbyte chip.
Then i want to burn the bootloader for the 512Kbyte chip. mriscoc have linked to one BL for the F401RE-84MHz, 512Kbyte

-You changed you small chip to the bigger one(no problem i can do that), but how about the BL?
-When you compile the mriscoc fw then you must have made some special change to match the new F401RE and BL?
-Some hints on what you have done?

**Reply by Crazy-Charles** [reply]
As the bootloader mriscoc linked to is for the F401RE i would try that one, however as none were available when i upgraded mine, i extracted the 401RC bootloader (64K starting at address 0x08000000) and patched the value at 0x348E from 0x02 to 0x06 to enable it to erase all the memory blocks and then flashed it back on the new microcontroller at the same starting address 0x08000000.
It was built using the STM32F401RE_creality (512K) environment option on the Marlin Auto Build extension, I think that is a standard option which mriscoc added a long time ago, no other changes were required.

Hope that helps

**Comment by Witska** [comment]
I replace STM32F401RCT6 to STM32F401RET6 on my S1 board. In STM32f401RCTx_BootLoader_v1.bin bootloader the value at 0x348E i changed from 0x02 to 0x06 and everything works.
[STM32f401RET6_BootLoader_E3S1.zip](https://github.com/mriscoc/Ender3V2S1/files/14736150/STM32f401RET6_BootLoader_E3S1.zip)

**Reply by Shaische** [reply]
@Witska If you don't mind helping, what address do you flash the boot loader and firmware respectively? I flashed boot loader at 0x08000000 and verified the flash successfully. Then flashed firmware at 0x08010000 and verified successfully via STM32 Cube Programmer. But when I go back to verify boot loader again, I get: "Error: Data mismatch found at address 0x08000000 (byte = 0x00 instead of 0x88)"

I'm trying to understand what I'm doing wrong here because I'm unable to connect via USB, it doesn't show up as a port, which tells me it's likely that the firmware isn't running to create the port? Any help appreciated as I'm new to STM32 SWD programming!

**Reply by Crazy-Charles** [reply]
You dont flash the firmware via the programmer as this erases the bootloader, just flash the bootloader at 0x08000000 and load the printer firmware using the memory card.

**Comment by bzyk02** [comment]
Hello! I bought an Ender 3s1 with Klipper installed and I have the same problem. I can't upload Marlin. I tried to upload the bootloader via STlinkV2 but I don't know if I'm doing it right. Can you help me? I've been struggling with this for a few days now and nothing.

Thx. Lukas 😁✌🏻

![20240914_230320](https://github.com/user-attachments/assets/82e44d00-4a99-4061-8220-43810daee56c)

**Comment by PolitehPetra** [comment]
Even manufacturers don't want to share the bootloader for their board with me. Can anyone tell me how to get around this? I have a programmer. I understand that each bootloader has its own rules for the firmware file, its location and name. I initially had the Klipper firmware and I need to somehow put it into the board, I can't do it directly from the programmer, since there is some special format from the Klipper itself. What bootloader can I install and how to name the firmware file for the STM32F401RCT6 microcontroller?
